spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java

```diff
@@ -70,29 +72,44 @@
 
 	private final AtomicInteger activeCount = new AtomicInteger();
 
+	@Override
 	public boolean onMessage(Message<?> message) {
-		if (this.shuttingDown) {
-			if (logger.isInfoEnabled()) {
-				logger.info("Inbound message ignored; shutting down; " + message.toString());
-			}
-		}
-		else {
-			if (message instanceof ErrorMessage) {
-				/*
-				 * Socket errors are sent here so they can be conveyed to any waiting thread.
-				 * There's not one here; simply ignore.
-				 */
-				return false;
+		boolean isErrorMessage = message instanceof ErrorMessage;
+		try {
+			if (this.shuttingDown) {
+				if (logger.isInfoEnabled()) {
+					logger.info("Inbound message ignored; shutting down; " + message.toString());
+				}
 			}
-			this.activeCount.incrementAndGet();
-			try {
-				return doOnMessage(message);
+			else {
+				if (isErrorMessage) {
+					/*
+					 * Socket errors are sent here so they can be conveyed to any waiting thread.
+					 * There's not one here; simply ignore.
+					 */
+					return false;
+				}
+				this.activeCount.incrementAndGet();
+				try {
+					return doOnMessage(message);
+				}
+				finally {
+					this.activeCount.decrementAndGet();
+				}
 			}
-			finally {
-				this.activeCount.decrementAndGet();
+			return false;
+		}
+		finally {
+			String connectionId = (String) message.getHeaders().get(IpHeaders.CONNECTION_ID);
+			if (connectionId != null && !isErrorMessage && this.isSingleUse) {
+				if (this.serverConnectionFactory != null) {
+					this.serverConnectionFactory.closeConnection(connectionId);
+				}
+				else {
+					this.clientConnectionFactory.closeConnection(connectionId);
```

Comment:
It doesn't relate to this PR, but looks like we really should do refactoring polishing around these `server/client` variables. I don't see so much different in usage here. 
So, the final code may be more simpler. 
Like background task :smile: 
Seems for me we haven't reached any scheduled background tasks yet...